src/cmds/qstat.c

```diff
@@ -302,47 +361,163 @@ states(char *string, char *q, char *r, char *h, char *w, char *t, char *e, int l
+					if (*value == ESC_CHAR && *(value + 1) == *comma)
+						value++;
+					keyvalpair[i++] = *value++;
+				}
+				keyvalpair[i] = '\0';
+				key = strtok_r(keyvalpair, "=", &val);
+				if (add_json_node(JSON_VALUE, JSON_NULL, key, val) == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+				}
+				if (*value != '\0') {
+					i = 0;
+					value++;
+				}
+			}
+			if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL) == NULL) {
+				fprintf(stderr, "qstat: out of memory\n");
+				exit(1);
+			}
+			free(keyvalpair);
+		} else {
+			if (resource) {
+				if (prev_resc_name && strcmp(prev_resc_name, name) != 0) {
+					if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL)
+							== NULL) {
+						fprintf(stderr, "qstat: out of memory\n");
+						exit(1);
+					}
+					prev_resc_name = NULL;
+				}
+				if (prev_resc_name == NULL || strcmp(prev_resc_name, name) != 0) {
+					if (add_json_node(JSON_OBJECT, JSON_NULL, name, NULL)
+							== NULL) {
+						fprintf(stderr, "qstat: out of memory\n");
+						exit(1);
+					}
+					prev_resc_name = name;
+				}
+				if (add_json_node(JSON_VALUE, JSON_NULL, resource, value)
+						== NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
+				}
+			} else {
+				if (prev_resc_name) {
+					if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL)
+							== NULL) {
+						fprintf(stderr, "qstat: out of memory\n");
+						exit(1);
+					}
+					prev_resc_name = NULL;
+				}
+				if (add_json_node(JSON_VALUE, JSON_NULL, name, value) == NULL) {
+					fprintf(stderr, "qstat: out of memory\n");
+					exit(1);
 				}
 			}
 		}
-		if ((c = strtok((char *)0, comma)) != NULL) {
-			first = 0;
-			putchar(',');
+		break;
+
+	case FORMAT_DSV:
+		buf = escape_delimiter(value, delimiter, ESC_CHAR);
+		if (buf == NULL) {
+			fprintf(stderr, "qstat: out of memory\n");
+			exit(1);
+		}
+		buf = convert_feed_chars(buf);
+		if (buf == NULL) {
+			fprintf(stderr, "qstat: out of memory\n");
+			exit(1);
+		}
+		if (resource)
+			printf("%s.%s=%s", name, resource, buf);
+		else
+			printf("%s=%s", name, buf);
+		free(buf);
+		break;
+
+	default:
+		if (one_line) {
+			buf = convert_feed_chars(value);
+			if (buf == NULL) {
+				fprintf(stderr, "qstat: out of memory\n");
+				exit(1);
+			}
+			if (resource)
+				printf("    %s.%s = %s", name, resource, buf);
+			else
+				printf("    %s = %s", name, buf);
+			free(buf);
+		} else {
+			start = strlen(name) + 7; /* 4 spaces + ' = ' is 7 */
+			printf("    %s", name);
+			if (resource) {
+				start += strlen(resource) + 1; /* 1 for dot */
+				printf(".%s", resource);
+			}
+			printf(" = ");
+			buf = NULL;
+			buf = strtok(value, comma);
+			while (buf) {
+				if ((len = strlen(buf)) + start < CHAR_LINE_LIMIT) {
+					printf("%s", buf);
+					start += len;
+				} else {
+					if (!first) {
+						printf("\n\t");
+						start = 9; /* 2 tabs + 1 */
+					}
+					while (*buf) {
```

Comment:
we can avoid going through buf character-by-character by using better logic. One suggestion would be - 
1) store the whole name(.resource) = value in buf 
2) check for its length (taking into account the first 4 spaces) doesn't go beyond CHAR_LINE_LIMIT. 
        a) If it doesn't cross CHAR_LINE_LIMIT - display the string
        b) If it does in a while loop - 
                i) print only the CHAR_LINE_LIMIT characters using ".*s" in the printf
                ii) add the required tabs and spaces on the next line.
                iii) advance the pointer in the buf by CHAR_LINE_LIMIT.
                iv) break when we are done printing the string.